bundler/spec/bundler/gem_version_promoter_spec.rb

```diff
@@ -1,175 +1,162 @@
     end
 
-    def build_package(name, platforms, locked_specs, unlock)
-      Bundler::Resolver::Package.new(name, platforms, :locked_specs => locked_specs, :unlock => unlock)
+    def sorted_versions(candidates:, current:, name: "foo", locked: [])
+      gvp.sort_versions(
+        build_package(name, current, locked),
+        build_candidates(candidates)
+      ).flatten.map(&:version).map(&:to_s)
     end
 
-    # Rightmost (highest array index) in result is most preferred.
-    # Leftmost (lowest array index) in result is least preferred.
-    # `build_candidates` has all versions of gem in index.
-    # `build_spec` is the version currently in the .lock file.
-    #
-    # In default (not strict) mode, all versions in the index will
-    # be returned, allowing Bundler the best chance to resolve all
-    # dependencies, but sometimes resulting in upgrades that some
-    # would not consider conservative.
-    context "filter specs (strict) level patch" do
-      let(:gvp) do
-        Bundler::GemVersionPromoter.new.tap do |gvp|
-          gvp.level = :patch
-          gvp.strict = true
-        end
-      end
+    it "numerically sorts versions" do
+      versions = sorted_versions(:candidates => %w[1.7.7 1.7.8 1.7.9 1.7.15 1.8.0], :current => "1.7.8")
+      expect(versions).to eq %w[1.7.7 1.7.8 1.7.9 1.7.15 1.8.0]
+    end
 
-      it "when keeping build_spec, keep current, next release" do
-        res = gvp.sort_versions(
-          build_package("foo", [], build_spec_set("foo", "1.7.8"), []),
-          build_candidates(%w[1.7.8 1.7.9 1.8.0])
-        )
-        expect(versions(res)).to eq %w[1.7.8 1.7.9]
+    context "with no options" do
+      it "defaults to level=:major, strict=false, pre=false" do
+        versions = sorted_versions(:candidates => %w[0.2.0 0.3.0 0.3.1 0.9.0 1.0.0 2.0.1 2.1.0], :current => "0.3.0")
+        expect(versions).to eq %w[0.2.0 0.3.0 0.3.1 0.9.0 1.0.0 2.0.1 2.1.0]
       end
+    end
 
-      it "when unlocking prefer next release first" do
-        res = gvp.sort_versions(
-          build_package("foo", [], build_spec_set("foo", "1.7.8"), []),
-          build_candidates(%w[1.7.8 1.7.9 1.8.0])
-        )
-        expect(versions(res)).to eq %w[1.7.8 1.7.9]
-      end
+    context "when strict" do
+      before { gvp.strict = true }
 
-      it "when unlocking keep current when already at latest release" do
-        res = gvp.sort_versions(
-          build_package("foo", [], build_spec_set("foo", "1.7.9"), []),
-          build_candidates(%w[1.7.9 1.8.0 2.0.0])
-        )
-        expect(versions(res)).to eq %w[1.7.9]
-      end
-    end
+      context "when level is major" do
+        before { gvp.level = :major }
 
-    context "filter specs (strict) level minor" do
-      let(:gvp) do
-        Bundler::GemVersionPromoter.new.tap do |gvp|
-          gvp.level = :minor
-          gvp.strict = true
+        it "keeps downgrades" do
```

Comment:
I think this is what was surprising to me. The other levels remove downgrades, but this one (the default), does not. I think this complicates things for Bundler outside of the gem version promoter because it needs to explicit add lower bound requirements to dependencies and I think it may be to workaround this.

Nothing to fix in this PR, but I think we may be able to remove a bit of complexity but making `level=major` behave like the other levels.

Reply:
Ah, gotcha. Yeah that had tripped me up a couple times. I'm not currently at the computer, but I wonder if changing https://github.com/rubygems/rubygems/blob/3586d716d0f8b6707afdafab37a6eb26033fffd2/bundler/lib/bundler/gem_version_promoter.rb#L80 with
```diff
-       return specs if locked_version.nil? || major?
+       return specs if locked_version.nil?
```
would fix that. Happy to try it either on this branch or a follow up!